### ISSUE TITLE
Some PDF certificates generated, others not

### ISSUE THREAD

**Issue by enrarelloem** [iomad/iomad#1682]
@turf212 The problem is a bit tricky to describe since I cannot see any pattern.
Some issued certificates are downloadable by users, appear in dashboard/reports/completion report by course and show in the activity certificate in the respective course - other don't.
![Report - completion report summary - 100 percent no certificate](https://user-images.githubusercontent.com/39500812/159693681-37deefd7-05c5-40bf-9184-a3856c6cb499.PNG)

![Report - completion report summary - 100 percent with certificate](https://user-images.githubusercontent.com/39500812/159693830-7924ab73-83fa-48b9-be6b-1075413cf251.PNG)

In the course, when viewing the list of issued certicates, there is no PDF symbol and the link to the PDF is just a dot. This link produces an error message "Sorry, the requested file could not be found".
![Course - activity certificate all groups](https://user-images.githubusercontent.com/39500812/159694183-f71966c6-094f-4274-8d26-69d2dc696655.PNG)

I can't really say when the problem occurred first, but I guess between end of last year and January.

Update history of the system
2022-03-23 Iomad 3.9.11+ to 3.9.13 
2021-11-17 Iomad 3.9.10+ to 3.9.11+
2021-09-22 Iomad 3.9.7 to 3.9.10+

When I enrol participants to a course just containing the certificate activity (without any conditions), certificates are generated. 

Any idea to solve the problem would be very much appreciated! Thanks!

**Comment by enrarelloem** [comment]
The generation of certificate PDFs seems to be completely random.
Here are screenshots from one customer. All participants did the course on the same license.
Pass grade is 75% and better.
In the certificate activity, no PDFs are displayed for this specific customer (just little dots leading to an error page).

![01 Course - activity certificate by separate groups CustomerV](https://user-images.githubusercontent.com/39500812/163384819-0bf2a8f6-73f9-4593-9db2-4b8bef1b7145.PNG)

In the course completion report for the customer, there are only two certificates though more participants qualify for the certificate. The two PDFs in the report are not shown in the certificate activity (compare image above).

![02 Report - completion report summary - some with-out certificate](https://user-images.githubusercontent.com/39500812/163384948-0be2a0f5-c848-4142-a452-2d6ee36c36df.PNG)

**Comment by turf212** [comment]
This is likely exactly the same issue as for #1690 - Moodle shouldn't allow white space in shortnames for courses.  It's a core issue and there is a tracker item mentioned in the issue I've just mentioned.

TLDR - needs to be fixed in core Moodle.

**Comment by enrarelloem** [comment]
Derick,
thanks for looking into this.
However, why should the system generate some certificates in a course and others not (in exactly the same course)?
It doesn't generate certificates for some customers and for other customers only some certificates are generated.
If it were a problem with the shortname of the course, I would expect the generation of certificates to fail completely. Or am I mistaken here?

**Comment by turf212** [comment]
Ah - apologies - I hadn't appreciated that.

If a user doesn't have access to the certificate (through restricted access rules) then that can cause the certificate not to be generated.  That logic was put in a while back to allow for multiple, different, certificate instances depending on e.g. a job title profile field.  Could it be this which is causing the issue?

**Comment by enrarelloem** [comment]
It does not seem very plausible to me since regular learners are affected and no special , restricted access rule should be in place. I'll check it anyway.

Here is another strange example. It's two views for five participants of one company (all the same role as learners, no special rules defined).
a) List of certificates in the course. There are just little dots which are links of the form: https://our-domain.com/pluginfile.php/9008/mod_iomadcertificate/issue/1366/  The last number in the URL is changing from participant to participant.
![grafik](https://user-images.githubusercontent.com/39500812/182376795-982d947c-16f8-44ea-8024-e0779bae6e26.png)

b) Iomad Reports - Completion report by course
![grafik](https://user-images.githubusercontent.com/39500812/182377150-49b4ef09-462b-442c-b6e2-aafcfd4e4d9d.png)
In b) all participants have fulfilled the conditions for a certificate. Participants "Z... S..." e.g. reached 100%, but there is no link to a certificate. The participants "C G" and "P H" had a broken link in the course view a) but working links in this report.

I wish I could see any pattern...

**Comment by turf212** [comment]
The call is being made in the code in local/iomad_track/classes/observer.php in the [function public static function course_completed($event)](https://github.com/iomad/iomad/blob/9ad8e49eda91d1f4fe5a7182c0ec1ef5af8cc506/local/iomad_track/classes/observer.php#L184)  line 378 calls the record_certificates() function.  That function (line 152) checks if the user can see the certificate and, if not, doesn't do anything further.  This could happen If the user is being unenrolled from the course on completion or if the restrict access rules are not being matched.

So - you could put some tracing into the function (use error log() as you don't want any output) to track the completion is being recorded OK (looks that way in the image) and where in the record_certificate the process is failing.

In the meantime - as long as the users all still have access, you can regenerate all of the certificates in that course by going to Site administration and clicking on the Recreate stored certificates link.  You can choose a company and or a course and or a time frame (the page will take a bit of time to load if you have a lot of users/companies/courses etc).

**Comment by enrarelloem** [comment]
@turf212
Sorry, for not replying earlier. The problem still exists. I had hoped an update to 4.0.x could fix it. Not being a developer or software engineer, I tried to avoid the detective work in the logs... I'll try to follow your suggestions and come back to it.
Thanks Derick.

**Comment by shadyshrif** [comment]
I went through the same issue, and It was solved when I removed the white spaces from the course short name.

I also found that the certificates couldn't be shown in the IOMAD report until the course is completed, so you have to set the course completion from your course more -> course completion.
![Screenshot 2024-07-15 at 09 41 21](https://github.com/user-attachments/assets/d40d0efd-c144-466f-b73d-3f5e3d51c3d5)

Then find the activity condition section and select ANY selected activities to be completed and choose which activity would complete the course, In my case it is the certificate, so the user has to get their certificate first.
![Screenshot 2024-07-15 at 09 44 37](https://github.com/user-attachments/assets/1e472dc2-b860-43fd-8c9a-2d557a4ef505)


I hope this helps.

**Comment by turf212** [comment]
Shortnames in Moodle shouldn't have special characters :)
